machineset controller for gcp (using syncsets)

joelddiaz opened the pull request:
new controller implementing support for syncing MachineSets for GCP clusters using SyncSets

joelddiaz commented:
/test e2e-gcp

joelddiaz commented:
/test e2e

staebler commented on pkg/controller/syncmachineset/syncmachineset_controller_test.go:
```diff
@@ -0,0 +1,475 @@
+						Value: &runtime.RawExtension{Object: &gcpprovider.GCPMachineProviderSpec{
+							MachineType: machineType,
+						},
+						},
```
Combine this onto the previous line since the corresponding opening braces are on the same line.

staebler commented on pkg/controller/syncmachineset/syncmachineset_controller_test.go:
```diff
@@ -0,0 +1,475 @@
+		},
+	}
+}
+func TestNewActuator(t *testing.T) {
```
`TestNewMachineSetActuator`

staebler commented on pkg/controller/syncmachineset/gcpactuator.go:
```diff
@@ -0,0 +1,126 @@
+
+	// Filter to regions matching '.*<region>.*'
+	zoneFilter := fmt.Sprintf("region eq '.*%s.*'", region)
+	zoneList, err := client.ListComputeZones(gcpclient.ListComputeZonesOptions{Filter: zoneFilter})
```
Although I think it is highly unlikely, we still need to handle the possibility that the results will be paged.

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator.go:
```diff
@@ -0,0 +1,126 @@
+	zones := []string{}
+
+	// Filter to regions matching '.*<region>.*'
+	zoneFilter := fmt.Sprintf("region eq '.*%s.*'", region)
```
using the installer would get us away from our mock client, so i'll just go ahead and add the 'status eq UP' filter

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator.go:
```diff
@@ -0,0 +1,126 @@
+		logger.Debugf("using image with name %s for machine sets", result.Items[0].Name)
+		return result.Items[0].Name, nil
+	default:
+		msg := fmt.Sprintf("unexpected number of results when looking for GCP image with name starting with %s", infra)
```
an excellent question. what sequence of events would leave us with with multiple images when running the openshift-installer?

staebler replied:
I don't know. I don't think that the installer would generate more than just the one. But I am picturing a scenario where there are, say, RHEL nodes running from a different image whose name also starts with the infra ID prefix.

joelddiaz replied:
i could change the regex to be more like what the installer is using https://github.com/openshift/installer/blob/master/pkg/asset/machines/gcp/machines.go#L89
wdyt?

staebler replied:
Might as well. Although at that point is there any need to even query GCP since we know the name already?

joelddiaz replied:
i guess all we get is an error bubbled up that the image is missing out in the cloud. i'd rather not write out a known bad machineset in case the cluster happens to be limping along in whatever config it is currently in.

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator_test.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+func generateMachineSetName(poolName, zone string) string {
+	return fmt.Sprintf("%s-%s-%s", generateInfraID(), poolName[:1], zone)
```
it seems that gcp has quite short name fields and that's what we get back from the installer https://github.com/openshift/installer/blob/master/pkg/asset/machines/gcp/machinesets.go#L45

staebler replied:
Fair enough, although I don't know that we need to mimic that behavior in unit tests: We can use whatever names we like.

staebler commented on pkg/controller/syncmachineset/syncmachineset_controller.go:
```diff
@@ -0,0 +1,290 @@
+	}
+
+	// Generate the machinesets we need to sync.
+	machineSets, err := r.generateMachineSets(cd, logger)
```
For example, you can see this in the logs from the e2e tests.
```
time="2019-11-18T16:41:11Z" level=warning msg="failed to generate machinesets" clusterDeployment=hive-9b2dc778-0d82-4128-ab82-3e175bbeb420 controller=syncmachineset error="cannot create actuator for unsupported platform" namespace=cluster-test
```

joelddiaz replied:
i completely forgot to add this. now fixed.

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator_test.go:
```diff
@@ -0,0 +1,279 @@
+			mockCtrl := gomock.NewController(t)
+			gClient := test.mockGCPClient(mockCtrl)
+
+			ga := &GCPActuator{
```
yes. plumbing through the gomock.Eq() stuff forced me to do this :) , although it's going to be conditional setting of region b/c of the test case where spec.platform.gcp is nil.

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator_test.go:
```diff
@@ -0,0 +1,279 @@
+	}
+}
+
+func validateMachineSets(t *testing.T, mSets []*machineapi.MachineSet, expectedMS []machineSetInfo) {
```
yes, there are extra fields that we aren't setting/testing for (things like CanIPForward, whether or not the installer sets DeletionProtection, the ProjectID gets inserted in there as well).

joelddiaz commented on pkg/controller/syncmachineset/syncmachineset_controller.go:
```diff
@@ -0,0 +1,290 @@
+	// Get current SyncSet object
+	ss := &hivev1.SyncSet{}
+	ssName := fmt.Sprintf("%s-machinesets", cd.Name)
+	if err := r.Get(context.TODO(), types.NamespacedName{Namespace: cd.Namespace, Name: ssName}, ss); err != nil {
```
no reason, this was likely leftover copy-pasting from the original aws machine set syncing controller. will remove.

joelddiaz commented on pkg/controller/syncmachineset/syncmachineset_controller.go:
```diff
@@ -0,0 +1,290 @@
+
+func getGCPClientCredentials(kubeClient client.Client, cd *hivev1.ClusterDeployment, logger log.FieldLogger) (*corev1.Secret, error) {
+	gcpCreds := &corev1.Secret{}
+	credsKey := types.NamespacedName{Name: cd.Spec.PlatformSecrets.GCP.Credentials.Name,
```
looks like that wouldn't pass validation https://github.com/openshift/hive/blob/master/pkg/apis/hive/v1alpha1/validating-webhooks/clusterdeployment_validating_admission_hook.go#L271 , but i suppose it's better safe than sorry

staebler replied:
Yes, we cannot rely on validation. We need to program defensively.

joelddiaz commented on pkg/controller/syncmachineset/syncmachineset_controller_test.go:
```diff
@@ -0,0 +1,475 @@
+	for _, raw := range ss.Spec.Resources {
+		ms, ok := raw.Object.(*machineapi.MachineSet)
+		if ok {
+			//gcpProvider, err := gcpprovider.ProviderSpecFromRawExtension(ms.Spec.Template.Spec.ProviderSpec.Value)
```
yes, i believe i was discovering that when using the fakeClient we have to fetch things out of value.object directly (if that's how we populated the object in the controller).

joelddiaz commented on pkg/controller/syncmachineset/syncmachineset_controller_test.go:
```diff
@@ -0,0 +1,475 @@
+				},
+			},
+			validate: func(t *testing.T, c client.Client, expectedMS []createdMachineSetInfo, createdSS createdSyncSetInfo) {
+				cd := getClusterDeployment(c)
```
because this pre-dated when i added the more generic validation function :). will fix.

joelddiaz commented:
@staebler i think i've addressed all the feedback except for the outstanding question about what the best approach is for finding the imageID to use for a machineset

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator.go:
```diff
@@ -0,0 +1,126 @@
+package syncmachineset
+
```
i tried to move this and ended up with an import loop of pkg/controller/syncmachineset importing pkg/controller/syncmachineset/gcp, and pkg/conroller/syncmachineset/gcp importing pkg/controller/syncmachineset (so that we can do the Actuator = &GCPActuator{} compile-time check)

joelddiaz replied:
actually, nevermind. i just needed to separate out the actuator definition into its own file. updated PR to move the gcp actuator into its own directory/package.

staebler replied:
We have generally not taken the approach of separating into directories by platform. We separate into directories by function instead. I would rather see the gcp actuator live in the same directory as all of the other actuators.

staebler commented on pkg/controller/syncmachineset/gcpactuator_test.go:
```diff
@@ -181,35 +174,28 @@ func TestGCPAcuator(t *testing.T) {
 
-				gcpProvider, ok := ms.Spec.Template.Spec.ProviderSpec.Value.Object.(*gcpprovider.GCPMachineProviderSpec)
-				assert.True(t, ok, "failed to convert to gcpProviderSpec")
+		assert.Equal(t, expected.replicas, int64(*mSets[i].Spec.Replicas), "replica mismatch")
```
That is interesting that replicas in machineset is int32 while replicas in machine pool is int64.

joelddiaz replied:
interesting indeed [surprised emoji here]

staebler commented on pkg/controller/syncmachineset/syncmachineset_controller.go:
```diff
@@ -288,3 +280,10 @@ func getGCPClientCredentials(kubeClient client.Client, cd *hivev1.ClusterDeploym
 }
+
+func isPlatformSupported(cd *hivev1.ClusterDeployment) bool {
+	if cd.Spec.Platform.GCP != nil {
```
I know that there is an expectation of supporting more platforms, but for now...
```
return cd.Spec.Platform.GCP != nil
```

staebler commented on pkg/controller/syncmachineset/syncmachineset_controller.go:
```diff
@@ -133,6 +133,12 @@ func (r *ReconcileSyncMachineSet) Reconcile(request reconcile.Request) (reconcil
 	}
 
+	// Do not reconcile a ClusterDeployment for a platform that isn't supported
+	if !isPlatformSupported(cd) {
```
Can you add a unit test for this?

joelddiaz replied:
done

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator_test.go:
```diff
@@ -51,11 +50,9 @@ func TestGCPAcuator(t *testing.T) {
-				mockListComputeZones(client, []string{"zone1"})
-				return client
+			mockGCPClient: func(client *mockgcp.MockClient, cd *hivev1.ClusterDeployment) {
+				mockListComputeImage(client, []string{"testImage"}, cd.Status.InfraID)
```
for the current list of test cases yes, but i was trying to handle someone coming in later and providing a non-standard clusterDeployment for a new test case.
but if it's premature optimization, i can pull it back to just use the global defaults

staebler replied:
The non-standard clusterdeployment does not change what the infra ID or region would be. In other words, there is no point in testing a cd with different values since the code just passes it along with any manipulation.

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator_test.go:
```diff
@@ -152,11 +139,9 @@ func TestGCPAcuator(t *testing.T) {
-				return client
+			mockGCPClient: func(client *mockgcp.MockClient, cd *hivev1.ClusterDeployment) {
+				mockListComputeImage(client, []string{"imageA"}, cd.Status.InfraID)
+				mockListComputeZones(client, []string{"zoneA"}, "")
```
just trying to get the code far enough so that when we call into the installer's MachineSet() generation we can simulate a failure

staebler replied:
OK. And that is entirely separate from listing the images and listing the zones. So there is no need to add the separate wrinkle of an empty region.

joelddiaz replied:
the empty region was because previously i had a nil spec.platform.gcp, but i can get the failure by filling out both spec.platform.aws and gcp simultaneously (even though that isn't really valid it still does the trick).

staebler replied:
I don't understand this. The region passed to `ListComputeZones` does not come from the clusterdeployment. Why does the contents of spec.platform factor in at all?

joelddiaz replied:
region does come from the clusterDeployment now that we've remove region from the struct.

gyliu513 commented on pkg/controller/syncmachineset/gcp/gcpactuator.go:
```diff
@@ -0,0 +1,146 @@
+package gcp
+
+import (
```
Thanks for the migration, with the file in the folder of `gcp`, I think we can rename the file to `actuator.go`?

staebler commented on pkg/controller/syncmachineset/syncmachineset_controller.go:
```diff
@@ -0,0 +1,289 @@
+}
+
+func (r *ReconcileSyncMachineSet) syncMachineSets(cd *hivev1.ClusterDeployment, logger log.FieldLogger) error {
+	logger.Info("reconciling machine sets for cluster deployment")
```
The log entry includes the name of the cluster deployment in the "clusterDeployment" field.

joelddiaz replied:
yep, in the Reconcile() function it is set up initially

gyliu513 replied:
thanks, got it.

joelddiaz commented on pkg/controller/syncmachineset/gcpactuator_test.go:
```diff
@@ -166,12 +151,20 @@ func TestGCPAcuator(t *testing.T) {
 				logger: log.WithField("actuator", "gcpactuator"),
 			}
+			if test.clusterDeployment.Spec.Platform.GCP != nil {
+				ga.region = test.clusterDeployment.Spec.Platform.GCP.Region
```
i have no idea what i was thinking when i put the region in the actuator struct. fixed.

joelddiaz commented:
@staebler i brought the actuator back into a single package with the controller so that the hive repo is at least consistent with how we organize our actuators. we can revisit organizing actuators in a separate PR.
PTAL at the current PR. i think the outstanding bits are around fetching the image for the machineSet

joelddiaz commented:
/test e2e-gcp

staebler commented:
The state of the PR looks good (although I still think that the `mockGCPClient` functions do not need the clusterdeployment parameter). Is this PR ready to go or are there still issues with setting up the corresponding subnets for the machinesets? Are we able to enable [this test](https://github.com/openshift/hive/blob/6231266bee9ea721f2b0caff6c12c16820dede48/test/e2e/postinstall/machinesets/infra_test.go#L18) for GCP now?

joelddiaz commented:
> The state of the PR looks good (although I still think that the `mockGCPClient` functions do not need the clusterdeployment parameter). Is this PR ready to go or are there still issues with setting up the corresponding subnets for the machinesets? Are we able to enable [this test](https://github.com/openshift/hive/blob/6231266bee9ea721f2b0caff6c12c16820dede48/test/e2e/postinstall/machinesets/infra_test.go#L18) for GCP now?

You've convinced me about removing the clusterDeployment from those mock setup calls. I'll push an update in a moment.

As far as the subnet stuff, as the code currently stands it will only push down the MachineSets that the installer would have generated for each of the worker pools, but nothing in-cluster will handle creating any necessary subnets. I believe @abutcher was saying that the AWS machine set syncing code also behaves this way which means this controller is at least feature equivalent with how AWS presently behaves.
@abutcher can you confirm that the AWS machine set syncing also doesn't handle the creation of any required additional subnets when trying to define extra machinepools post-install?

abutcher commented:
@joelddiaz The AWS machineset code does not handle creating additional resources and we currently reuse the original worker pool's resources.

joelddiaz commented:
/test e2e-gcp

joelddiaz commented:
@staebler with #708 merged (limiting machine pools to length 1 and name 'worker'), are we good for this?
